### PR TITLE
Integration with elasticmem

### PR DESCRIPTION
**Pull request by pansicheng**
## Motivation

This PR implements dynamic scaling between different attention-type pools within the hybrid model in sglang, based on elasticmem.

______________________________________________________________________a-Flowchart (10)

## Benchmarking and Profiling

```bash

export SGLANG_ELASTIC_MEM_POOL=true
export SGLANG_RATIO=1.0
nohup python3 -m sglang.launch_server \
  --log-level debug \
  --model /home/t4/models/lvm-data/Llama-4-Scout-17B-16E-Instruct \
  --tp 2 \
  --attention-backend fa3 \
  --hybrid-kvcache-ratio ${SGLANG_RATIO} \
  --context-length 200000 \
  > nohup.emem.${SGLANG_ELASTIC_MEM_POOL}.ratio.${SGLANG_RATIO}.out 2>&1 \
  &

export SGLANG_ELASTIC_MEM_POOL=true
export SGLANG_RATIO=1.0
nohup python3 -m sglang.bench_serving --backend sglang \
  --dataset-name random --dataset-path /home/t4/models/lvm-data/ShareGPT_V3_unfiltered_cleaned_split.json \
  --num-prompts 1024 --random-input 1024 --random-output 1024 --random-range-ratio 1 \
  --max-concurrency 128 \
  > nohup.bench.${SGLANG_ELASTIC_MEM_POOL}.ratio.${SGLANG_RATIO}.out 2>&1 \
  &
```

image

- Horizontal axis: Each time step represents a log entry for Prefill/Decode.
- Vertical axis: As shown in the figure, it represents:
  - Token usage for different pools,
  - Real-time running requests,
  - Generation throughput during decoding (set to 0 during prefill).
- First row: Current static pool configuration with --hybrid-kvcache-ratio=0.6 (relatively balanced allocation between the full and swa pools).
- Second row: Current static pool configuration with --hybrid-kvcache-ratio=1.0 (most GPU memory allocated to the full pool).
- Third row: Elastic pool configuration with --hybrid-kvcache-ratio=1.0 (initial allocation favors the full pool but supports dynamic runtime adjustments).

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -1768,6 +1715,7 @@ def get_new_batch_prefill(self) -> Optional[ScheduleBatch]:
         if self.enable_lora:
             lora_set = set([req.lora_id for req in self.running_batch.reqs])
 
+        flag_no_token = False
```
Comment (yizhang2077) [inline python/sglang/srt/managers/scheduler.py]: need to be more detailed for flag name

File: scripts/emem/plot/main.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
+import os
+import re
+
+import matplotlib.pyplot as plt
```
Comment (yizhang2077) [inline scripts/emem/plot/main.py:6]: move `emem` dir under scripts

File: python/sglang/srt/mem_cache/elastic/elastic_allocator.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    def merge_and_sort_free(self):
+        if len(self.release_pages) > 0:
+            super().merge_and_sort_free()
+        else:
+            self.free_pages, _ = torch.sort(self.free_pages)
+
+    # TODO: a more efficient way
+    @override
+    def alloc(self, need_size: int):
+        self.merge_and_sort_free()
```
Comment (yizhang2077) [inline python/sglang/srt/mem_cache/elastic/elastic_allocator.py]: make it more efficient

File: python/sglang/srt/mem_cache/elastic/elastic_allocator.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        return super().alloc(need_size)
+
+    @override
+    def can_unmap(self) -> bool:
+        if self.token_usage() > 0.9:
+            return False
+
+        self.evict(self.evictable_size())
```
Comment (yizhang2077) [inline python/sglang/srt/mem_cache/elastic/elastic_allocator.py]: does `can_unmap` need to `evict` and `merge_and_sort`, since both seems to be time consuming

File: scripts/emem/plot/main.py (hunk 2 above)
Comment (pansicheng) [reply scripts/emem/plot/main.py:6]: done

File: python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+from abc import ABC, abstractmethod
+from typing import Tuple
+
+import torch
+
+from sglang.srt.utils import get_bool_env_var, get_int_env_var
+
+logger = logging.getLogger(__name__)
+
+use_elasticmem = get_bool_env_var("SGLANG_ELASTIC_MEM_POOL", "false")
```
Comment (yizhang2077) [inline python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py:29]: disable it when `enable-memory-saver` is true

File: python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                min_token_usage = allocator.token_usage()
+                min_token_usage_allocator = allocator
+
+            if allocator.can_unmap() and (
+                (unmap_candidate is None)
+                or (allocator.token_usage() < unmap_candidate.token_usage())
+            ):
+                unmap_candidate = allocator
+
+        if unmap_candidate is None and min_token_usage < 0.9:
```
Comment (yizhang2077) [inline python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py]: Could we make constant like `0.9` or `0.1` configurable? or we need a name?

File: python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+            and map_candidate != unmap_candidate
+        ):
+            logger.info(
+                "ElasticMempoolOrchestrator try_resize "
+                f"{map_candidate.token_usage()=}, "
+                f"{unmap_candidate.token_usage()=}"
+            )
+            self.do_resize(map_candidate, unmap_candidate)
+
+    def do_resize(
```
Comment (yizhang2077) [inline python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py]: use `_do_resize`?

File: python/sglang/srt/managers/scheduler.py (hunk 1 above)
Comment (pansicheng) [reply python/sglang/srt/managers/scheduler.py]: removed

File: python/sglang/srt/mem_cache/elastic/elastic_allocator.py (hunk 3 above)
Comment (pansicheng) [reply python/sglang/srt/mem_cache/elastic/elastic_allocator.py]: now we sort only during defragmentation

File: python/sglang/srt/mem_cache/elastic/elastic_allocator.py (hunk 4 above)
Comment (pansicheng) [reply python/sglang/srt/mem_cache/elastic/elastic_allocator.py]: Now can_unmap skips eviction and merge_and_sort, using an unused_pages tensor to track consecutive tail pages

File: python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py (hunk 6 above)
Comment (pansicheng) [reply python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py]: made them configurable via environment variables

File: python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py (hunk 7 above)
Comment (pansicheng) [reply python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py]: done

**Comment by pansicheng** [comment]
> Nice. I see how we try to improve max running batch size with this. In parallel, do we target to improve prefix cache hit rate with this as well by analyzing which of swa or full causing cache hit miss?

@hanming-lu No problem. The current PR focuses on balancing pool usage to maximize batch size when some pools near capacity. Next, we’ll monitor cache hit rates per pool and optimize scaling strategies to boost hit rates under balanced loads. Metrics and adaptive scaling will need further design, let’s tackle this next!

File: python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py (hunk 5 above)
Comment (pansicheng) [reply python/sglang/srt/mem_cache/elastic/elasticmem_orchestrator.py:29]: done

**Comment by ZelinMa557** [comment]
hi, will this feature support GDN models and mamba models?

**Comment by pansicheng** [comment]
> hi, will this feature support GDN models and mamba models?

There’s a PR for Qwen3-Next support here: https://github.com/sgl-project/sglang/pull/14597. I’ll try to move it forward as soon as possible.

**Comment by ZelinMa557** [comment]
> > hi, will this feature support GDN models and mamba models?
> 
> There’s a PR for Qwen3-Next support here: #14597. I’ll try to move it forward as soon as possible.

Thanks for your reply, I'm very interested in supporting dynamic memory pool for mamba/GDN models, is there anything I can help with?

**Comment by kevincheng2** [comment]
Could you please advise on how to evict swa cache in a hybrid cache scenario where swa has more cache (mostly outside the window), to ensure that the ratio between full and swa pool sizes is appropriate?
